facebook/draft-js
# Support rendering Draft.js into iframes (nested browsing contexts)

### Summary

Removes Draft’s dependency on the global `window` and `document` objects and makes it work across nested browsing contexts (iframes) by using the document and window relative to the browsing context in which Draft is being used. React itself makes similar assumptions about being able to use the global `window` and `document` objects, as documented in facebook/react#427, so this PR depends on facebook/react#9184. All together, these changes make it possible to render Draft.js inside an iframe, which in turn makes it possible to have a UX where the page includes multiple visible selections, like [in this demo](http://codepen.io/acusti/pen/RGEJZE?editors=0011). In that demo, I used built versions of React and Draft.js which include all the changes necessary to make it work across nested browsing contexts.

Fixes #527

### Test Plan

All of the unit tests are passing. For manual testing, I’ve focused on verifying the behavior of the editor when dealing with selections. That includes:

- Select a portion of text and start typing to replace it
- Select a portion of text and hit delete to remove it
- Select a portion of text and paste in content to replace it
- Select a portion of text and copy it, then paste it somewhere
- Select a portion of text and cut it, then paste it somewhere

I then:

- Repeated the steps above with a collapsed selection (for the insertion step)
- Repeated the steps above with a selection that spans multiple blocks

Finally, I manually tested all of those behaviors with:

- A regular editor implementation where the editor is inside the same browsing context as the rest of the React app
- An implementation where the editor is rendered into an iframe within the parent browsing context where React is running, like with the codepen demo I linked to above

In order to make it easy to try and test these behaviors, I created installable branches of fbjs and React that include the `lib/` files to make it installable and usable via npm as well as the built files to be usable directly in the browser.

### Try it out via npm

In your package.json, use:

```js
    "draft-js": "Brandcast/draft-js#f9affa3",
    "fbjs": "Brandcast/fbjs-built#97b8e54",
    "react": "Brandcast/react#9a81d1a",
```

### Try it out in the browser

Include:
```html
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react.js"></script>
    <script src="https://rawgit.com/Brandcast/react/nested-browsing-contexts-2-built/build/react-dom.js"></script>
    <script src="https://cdn.jsdelivr.net/immutable.js/latest/immutable.min.js"></script>
    <script src="https://cdn.jsdelivr.net/es6.shim/0.35.1/es6-shim.min.js"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/babel-core/5.8.34/browser.js"></script>
    <script src="https://rawgit.com/Brandcast/draft-js/iframe-compatibility-built/dist/Draft.js"></script>
```

**Rendering into an iframe**

To know more about rendering into an iframe, see [this explanation from @ryanseddon](https://developer.zendesk.com/blog/rendering-to-iframes-in-react). Also, he turned his work into the [React Frame Component](https://github.com/ryanseddon/react-frame-component), which makes it trivial to try out. And to understand use cases and the utility of enabling this functionality, the major benefits are:

1. Enabling multiple visible selections (see [the demo](http://codepen.io/acusti/pen/RGEJZE?editors=0011) I mentioned above).
2. Proper CSS encapsulation (e.g. render a preview of something with custom CSS without worrying about loading those styles into the app `window`).
